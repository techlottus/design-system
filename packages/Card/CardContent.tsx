import * as React from 'react';

interface CardContentProps {
  children: any;
  className: string;
}

const CardContent: React.SFC<CardContentProps> = ({ children, className }: CardContentProps) => (
  <div className={`${className}`}>{children}</div>
);

CardContent.defaultProps = {
  className: 'px-5 py-5',
};

export default CardContent;
