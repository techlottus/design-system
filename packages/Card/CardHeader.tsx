import * as React from 'react';

interface CardHeaderProps {
  children: any;
}

const CardHeader: React.SFC<CardHeaderProps> = ({ children }: CardHeaderProps) => (
  <React.Fragment>{children}</React.Fragment>
);

export default CardHeader;
