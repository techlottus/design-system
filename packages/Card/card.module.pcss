.card {
  @apply bg-white rounded overflow-hidden shadow-up-box;

  &Bordered {
    @apply border-t-8 rounded-card;
  }
}