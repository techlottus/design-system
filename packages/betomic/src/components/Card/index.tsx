import React, { Children } from "react";
import cn from "classnames";

type CardProps = {
  children: React.ReactNode;
  className?: string;
};

type CardComposition = {
  Header: React.FC<CardProps>;
  Body: React.FC<CardProps>;
  Footer: React.FC<CardProps>;
};

const CardChild: React.FC<CardProps> = ({ children, className }) => {
  return (
    <div className={className}>{children}</div>
  );
};

const Card: React.FC<CardProps> & CardComposition = (props) => {

  const { children, className } = props;

  const cardClasses = cn("flex flex-col p-4 bg-white rounded shadow-sm hover:shadow", className);

  return (
    <div className={cardClasses}>
      {Children.map(children, child => {
        if (React.isValidElement(child)) {
          const newChild = React.cloneElement(child);
          return newChild;
        }
        return;
      })}
    </div>
  );
};

Card.Header = CardChild;
Card.Body = CardChild;
Card.Footer = CardChild;

export default Card;