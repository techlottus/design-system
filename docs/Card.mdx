---
name: Card
menu: Components
---

import { Playground, Props } from 'docz'
import Card from '../packages/Card'
import CardHeader from '../packages/Card/CardHeader'
import CardContent from '../packages/Card/CardContent'
import CardFooter from '../packages/Card/CardFooter'
import Button from '../packages/Button';

# Card

Card component docs 📄

<Props of={Card} />

## Basic

<Playground>
  <Card>
    <CardHeader>
      <img
      className="rounded-lg shadow-xl h-56 w-full object-fit object-center"
      src="https://bedu.org/wp-content/uploads/Bedu07_ux_frontend.png" />
    </CardHeader>
    <CardContent>
      <div className="text-xl font-montserrat font-bold text-surface-900 mb-2">
        Fullstack JavaScript
      </div>
      <p className="font-base text-surface-700">
        Aprende a desarrollar web apps de forma fluida y autónoma con el ecosistema tecnológico más popular.
      </p>
    </CardContent>
    <CardFooter>
      <Button>Ver más</Button>
    </CardFooter>
  </Card>
</Playground>

## With Border

<Playground>
  <Card bordered>
    <CardHeader>
      <img
      className="rounded-lg shadow-xl h-56 w-full object-fit object-center"
      src="https://bedu.org/wp-content/uploads/Bedu07_ux_frontend.png" />
    </CardHeader>
    <CardContent>
      <div className="text-xl font-montserrat font-bold text-surface-900 mb-2">
        Fullstack JavaScript
      </div>
      <p className="font-base text-surface-700">
        Aprende a desarrollar web apps de forma fluida y autónoma con el ecosistema tecnológico más popular.
      </p>
    </CardContent>
    <CardFooter>
      <Button>Ver más</Button>
    </CardFooter>
  </Card>
</Playground>

## Without Header/Footer

<Playground>
  <Card bordered variant="english" variantAs="500">  
    <CardContent>
      <div className="text-xl font-montserrat font-bold text-surface-900 mb-2">
        Fullstack JavaScript
      </div>
      <p className="font-base text-surface-700">
        Aprende a desarrollar web apps de forma fluida y autónoma con el ecosistema tecnológico más popular.
      </p>
    </CardContent>
  </Card>
</Playground>