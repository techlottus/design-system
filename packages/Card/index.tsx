import * as React from 'react';
import classNames from '../common/classnames';
import { variantIntensity, variantName } from '../common/constants';
import CardHeader from './CardHeader';
import CardContent from './CardContent';
import CardFooter from './CardFooter';
import styles from './card.module.pcss';

export interface ICardProps {
  children?: React.ReactNode;
  bordered?: boolean;
  variant?: string;
  variantAs?: string;
}

class Card extends React.Component<ICardProps> {
  public static Header = CardHeader;
  public static Content = CardContent;
  public static Footer = CardFooter;
  static defaultProps: { bordered: boolean; variant: string; variantAs: string; };

  public render() {
    const { children, bordered, variant, variantAs } = this.props;
    const classes = classNames(
      styles.cardBordered,
      `border-${variantName[variant]}-${variantIntensity[variantAs]}`,
    );

    return(
      <div className={styles.card}>
        {bordered && <div className={classes} />}
        {children}
      </div>
    );
  }
}

Card.defaultProps = {
  bordered: false,
  variant: 'surface',
  variantAs: '500',
};

export default Card;
