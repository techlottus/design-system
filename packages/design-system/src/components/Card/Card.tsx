import cn from "classnames";
import TextLink from "../TextLink";
import Aspect from "../Aspect";
import { myhref } from "../helpers/myrefHelper";
import Heading from "../Heading";
import { CardType, DisplayType } from "../Types/Card.types";

const types: DisplayType = {
  horizontal: "w-3/5",
  vertical: "h-full",
};

const display: DisplayType = {
  horizontal: "flex-row",
  vertical: "flex-col",
};
const classesContent: string = cn(
  "flex flex-col",
  "relative",
  "justify-between",
  "p-4"
);
const classesSubTitle: string = cn(
  "flex",
  "pb-2",
  "font-bold font-principal text-neutral-500",
  " md:text-sm sm:text-xs"
);
const classesLink: string = cn(
  "flex",
  "font-bold font-principal",
  "justify-end items-center",
  "lg:text-base md:text-sm sm:text-xs"
);

const classesText: string = cn(
  "pb-2",
  "text-neutral-500",
  "font-nunito",
  " md:text-sm sm:text-xs"
);
const classesCard: string = cn(
  "flex flex-1",
  "relative",
  "rounded border border-neutral-300",
  "overflow-hidden h-full"
);

const Card: React.FC<CardType> = (props: CardType) => {
  const {
    imageUrl,
    content,
    subtitle,
    title,
    textLink,
    type = "horizontal",
    ImageAspectRatio = "2/1",
    className = "",
  } = props;
  const classText: string = cn(classesText, { ["hidden"]: !content });
  const classContent: string = cn(classesContent, { [types[type]]: type });
  const classSubTitle: string = cn(classesSubTitle, { ["hidden"]: !subtitle });
  const classLink: string = cn(classesLink, { ["hidden"]: !textLink });
  const classTitle: string = cn("text-neutral-800 pb-2", {
    ["hidden"]: !title,
  });
  const classCard: string = cn(classesCard, className, {
    [display?.[type]]: type,
    ["group hover:shadow-lg  cursor-pointer"]: textLink?.href,
  });

  const handleOnClick = (e: Event) => {
    if (textLink?.disabled) {
      e.stopPropagation();
    } else if (textLink?.href) {
      myhref(textLink?.href);
    } else if (textLink?.onClick) {
      textLink?.onClick();
    }
  };

  return (
    <div
      className={classCard}
      onClick={(e: any) => {
        handleOnClick(e);
      }}
    >
      {type === "horizontal" ? (
        <div id="image" className="w-2/5 h-auto ">
          <img
            className="w-full h-full"
            src={imageUrl}
            style={{ objectFit: "cover", objectPosition: "center" }}
          />
        </div>
      ) : (
        <Aspect ratio={ImageAspectRatio}>
          <div id="image" className="w-full h-full ">
            <img
              className="w-full h-full object-cover object-center"
              src={imageUrl}
            />
          </div>
        </Aspect>
      )}
      <div className={classContent}>
        <div>
          <div id="subtitle" className={classSubTitle}>
            <span>{subtitle}</span>
          </div>
          <div id="title" className={classTitle}>
            <Heading title={title} variant="h-6" font="secondary" />
          </div>
          <div id="content" className={classText}>
            <span dangerouslySetInnerHTML={{ __html: content }} />
          </div>
        </div>
        <div className={classLink}>
          <TextLink
            className="group-hover:underline text-sm"
            iconName="arrow_forward_ios"
            {...textLink}
          />
        </div>
      </div>
    </div>
  );
};

export default Card;
