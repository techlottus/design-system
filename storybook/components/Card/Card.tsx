import cn from "classnames";
import TextLink from "../TextLink";
import Aspect from "../Aspect";
import { myhref } from "../helpers/myrefHelper";
import Heading from "../Heading";
import { CardType } from "../Types/Card.types";
import React from "react";
const Card: React.FC<CardType> = (props: CardType) => {
  const {
    imageUrl,
    content,
    subtitle,
    title,
    textLink,
    orientation = "horizontal",
    imageAspectRatio = "2/1",
    className = "",
  } = props;
  /** classes for text in textlink */
  const classText: string = cn(
    "pb-2 text-neutral-500 font-nunito md:text-sm sm:text-xs",
    { ["hidden"]: !content }
  );
  const classContent: string = cn(
    "flex flex-col relative justify-between p-4",
    {
      ["w-3/5"]: orientation === "horizontal",
      ["h-full"]: orientation === "vertical",
    }
  );
  const classSubTitle: string = cn(
    "flex pb-2 font-bold font-principal text-neutral-500 md:text-sm sm:text-xs",
    { ["hidden"]: !subtitle }
  );
  const classLink: string = cn(
    "flex font-bold font-principal justify-end items-center lg:text-base md:text-sm sm:text-xs",
    { ["hidden"]: !textLink }
  );
  const classTitle: string = cn("text-neutral-800 pb-2", {
    ["hidden"]: !title,
  });
  const classCard: string = cn(
    "flex flex-1 relative rounded border border-neutral-300 overflow-hidden h-full",
    className,
    {
      ["flex-row"]: orientation === "horizontal",
      ["flex-col"]: orientation === "vertical",
      ["group hover:shadow-lg  cursor-pointer"]: textLink?.href,
    }
  );

  const handleOnClick = (e: Event) => {
    if (textLink?.disabled) {
      e.stopPropagation();
    } else if (textLink?.href) {
      myhref(textLink?.href);
    } else if (textLink?.onClick) {
      textLink?.onClick();
    }
  };

  return (
    <div
      className={classCard}
      onClick={(e: any) => {
        handleOnClick(e);
      }}
    >
      {
        /** when horizontal image grows with content */ orientation ===
        "horizontal" ? (
          <div id="image" className="w-2/5 h-auto ">
            <img
              className="w-full h-full object-cover object-center"
              src={imageUrl}
            />
          </div>
        ) : (
          /** when is vertical, image receive an aspect ratio */ <Aspect
            ratio={imageAspectRatio}
          >
            <div id="image" className="w-full h-full ">
              <img
                className="w-full h-full object-cover object-center"
                src={imageUrl}
              />
            </div>
          </Aspect>
        )
      }
      <div className={classContent}>
        <div>
          <div id="subtitle" className={classSubTitle}>
            <span>{subtitle}</span>
          </div>
          <div id="title" className={classTitle}>
            <Heading title={title} variant="h-6" font="secondary" />
          </div>
          <div id="content" className={classText}>
            <span dangerouslySetInnerHTML={{ __html: content }} />
          </div>
        </div>
        <div className={classLink}>
          <TextLink
            className="group-hover:underline text-sm"
            iconName="arrow_forward_ios"
            {...textLink}
          />
        </div>
      </div>
    </div>
  );
};

export default Card;
