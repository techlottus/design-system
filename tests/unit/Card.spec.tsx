import * as React from 'react';
import { shallow } from 'enzyme';
import Card from '../../packages/Card';
import CardHeader from '../../packages/Card/CardHeader';

describe('Card.tsx', () => {
  it('Should match the snapshot', () => {
    const vm = shallow(<Card><h1>New Card</h1></Card>);
    expect(vm).toMatchSnapshot();
  });

  it('The component Card has the correct classname', () => {
    const vm = shallow(<Card><h1>New Card</h1></Card>);
    expect(vm.props().className).toBe('max-w-sm rounded overflow-hidden shadow-up-box mx-auto');
  });

  it('The prop children accepts a CardHeader component as children', () => {
    const vm = shallow(<Card><CardHeader><h1>New Card with header</h1></CardHeader></Card>);
    expect(typeof vm.props().children).toBe('object');
  });
});
