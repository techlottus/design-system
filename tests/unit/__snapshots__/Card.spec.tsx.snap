// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card.tsx Should match the snapshot 1`] = `
<div
  className="max-w-sm rounded overflow-hidden shadow-up-box mx-auto"
>
  <h1>
    New Card
  </h1>
</div>
`;
